main/lsp/requests/hover.cc

```diff
@@ -48,20 +75,38 @@ LSPResult LSPLoop::handleTextDocumentHover(unique_ptr<core::GlobalState> gs, con
         }
 
         auto resp = move(queryResponses[0]);
+
+        optional<string> documentation = nullopt;
+        if (
+                resp->isIdent() || 
+                resp->isLiteral() || 
+                resp->isConstant() || 
+                resp->isDefinition()) {
+            auto loc = resp->getTypeAndOrigins().origins[0];
+            documentation = findDocumentation(loc.file().data(*gs).source(), loc.beginPos());
+        }
+
         if (auto sendResp = resp->isSend()) {
             auto retType = sendResp->dispatchResult->returnType;
+            // TODO main.method is <none> when calling .new on a class.
+            auto loc = sendResp->dispatchResult->main.method.data(*gs)->loc();
+            documentation = findDocumentation(loc.file().data(*gs).source(), loc.beginPos());
             auto &constraint = sendResp->dispatchResult->main.constr;
             if (constraint) {
                 retType = core::Types::instantiate(core::Context(*gs, core::Symbols::root()), retType, *constraint);
             }
             response->result = make_unique<Hover>(formatRubyCode(
-                clientHoverMarkupKind, methodSignatureString(*gs, retType, *sendResp->dispatchResult, constraint)));
+                clientHoverMarkupKind, methodSignatureString(*gs, retType, *sendResp->dispatchResult, constraint),
+                documentation));
         } else if (auto defResp = resp->isDefinition()) {
             response->result = make_unique<Hover>(formatRubyCode(
-                clientHoverMarkupKind, methodDetail(*gs, defResp->symbol, nullptr, defResp->retType.type, nullptr)));
+                clientHoverMarkupKind, methodDetail(*gs, defResp->symbol, nullptr, defResp->retType.type, nullptr),
+                documentation));
         } else if (auto constResp = resp->isConstant()) {
             const auto &data = constResp->symbol.data(*gs);
             auto type = constResp->retType.type;
+            // TOOD This constResp only has a reference to where the class is referred to. Can't seem to get to the
+            // definition directly using the properties from the response.
```

Comment:
I think you fixed this above by using `getTypeAndOrigin`.

Reply:
Oh, yeah. 64998a434